contributors/design-proposals/daemonset-update.md

```diff
@@ -0,0 +1,268 @@
+
+Deleting a DaemonSet (with cascading) will delete all its pods and PodTemplates. 
+
+
+## DaemonSet Strategies
+
+DaemonSetStrategy specifies how the new daemon pods should replace existing ones.
+To begin with, we will support 2 types:
+
+* On delete: Do nothing, until existing daemon pods are killed (for backward
+  compatibility).
+  - Other alternative names: No-op, External
+* Rolling update: We gradually kill existing ones while creating the new one.
+
+
+## Tests
+
+- Updating a RollingUpdate DaemonSet will trigger updates to its daemon pods. 
+- Updating an OnDelete DaemonSet will not trigger updates, until the pods are
+  killed. 
+- Users can use node labels to choose which nodes this DaemonSet should target.
+  DaemonSet updates only affect pods on those nodes.
+  - For example, some nodes may be running manifest pods, and other nodes will
+    be running daemon pods 
+- DaemonSet should support at most one daemon pod per node guarantee.
+  - Adding or deleting nodes won't break that.
+- Users should be able to specify acceptable downtime of their daemon pods, and
+  DaemonSet updates should respect that. 
+- DaemonSets can be updated while already being updated (i.e. rollover updates)
+- Broken rollout can be rolled back (by applying old config)
+
+
+## Future Plans
+
+In the future, we may:
+
+- Implement at-most-one and/or at-least-one guarantees for DaemonSets (i.e. at
+  most/at least one pod from a DaemonSet can exist on a node at any time)
+- Support use cases where uptime is critical for each pod of a DaemonSet during
+  an upgrade
+  - One approach is to use dummy pods to pre-pull images to reduce down time 
+- Support use cases that each DaemonSet pod can still fit on the node after
+  being updated (unless it becomes larger than the node). Some possible
+  approaches include:
+  - Make DaemonSet pods (daemons) have higher priority than non-daemons, and
+    kubelet will evict pods with lower priority to make room for higher priority
+    ones 
+  - The DaemonSet controller will evict pods when daemons can't fit on the node
+  - The DaemonSet controller will cordon the node before upgrading the daemon on
+    it, and uncordon the node once it's done
+- Support use cases that require the node to be drained before the daemons on it 
+  can updated (e.g. logging daemons)
+  - The DaemonSet controller will drain the node before upgrading the daemon on
+    it, and uncordon the node once it's done 
+- Make DaemonSets admin-only resources (admon = people who manage nodes). Some
+  possible approaches include:
+  - Remove namespace from DaemonSets (DaemonSets become node-level resources)
+  - Modify RBAC bootstrappolicy to make DaemonSets admin-only 
+  - Delegation or impersonation 
+- Support more DaemonSet update strategies 
+- Support rolling back DaemonSets 
+  - `kubectl rollout undo daemonset/<DaemonSet-Name>` to roll back 
+- Allow user-defined DaemonSet unique label key
+- Support pausing DaemonSets
```

Comment:
s/pausing DaemonSets/pausing DaemonSet rolling update/ is what I assume you meant?